libcloud/common/gandi_live.py

```diff
@@ -0,0 +1,250 @@
+#        "description": "\"FAKEUUID\" is not a UUID"
+#      }
+#   ]
+# }
+class InvalidRequestError(GandiLiveBaseError):
+    pass
+
+
+# Examples:
+# {
+#   "code": 409,
+#   "message": "Zone Testing already exists",
+#   "object": "HTTPConflict",
+#   "cause": "Duplicate Entry"
+# }
+# {
+#   "code": 409,
+#   "message": "The domain example.org already exists",
+#   "object": "HTTPConflict",
+#   "cause": "Duplicate Entry"
+# }
+# {
+#   "code": 409,
+#   "message": "This zone is still used by 1 domains",
+#   "object": "HTTPConflict",
+#   "cause": "In use"
+# }
+class ResourceConflictError(GandiLiveBaseError):
+    pass
+
+
+class GandiLiveResponse(JsonResponse):
+    """
+    A Base Gandi Live Response class to derive from.
+    """
+
+    def success(self):
+        """
+        Determine if our request was successful.
+
+        For the Gandi Live response class, tag all responses as successful and
+        raise appropriate Exceptions from parse_body.
+
+        :return: C{True}
+        """
+
+        return True
+
+    def parse_body(self):
+        """
+        Parse the JSON response body, or raise exceptions as appropriate.
+
+        :return:  JSON dictionary
+        :rtype:   ``dict``
+        """
+        if len(self.body) == 0 and not self.parse_zero_length_body:
+            return self.body
+
+        json_error = False
+        try:
+            body = json.loads(self.body)
+        except Exception:
+            # If there is both a JSON parsing error and an unsuccessful http
+            # response (like a 404), we want to raise the http error and not
+            # the JSON one, so don't raise JsonParseError here.
+            body = self.body
+            json_error = True
+
+        # Service does not appear to return HTTP 202 Accepted for anything.
+        valid_http_codes = [
+            httplib.OK,
+            httplib.CREATED,
+            httplib.NO_CONTENT
+        ]
+        if self.status in valid_http_codes:
+            if json_error:
+                raise JsonParseError(body, self.status)
+            else:
+                return body
+        elif self.status == httplib.NOT_FOUND:
+            if (not json_error) and ('cause' in body):
+                message = self._get_error(body)
+            else:
+                message = body
+            raise ResourceNotFoundError(message, self.status)
+        elif self.status == httplib.BAD_REQUEST:
+            if (not json_error) and ('cause' in body):
+                message = self._get_error(body)
+            else:
+                message = body
+            raise InvalidRequestError(message, self.status)
+        elif self.status == httplib.CONFLICT:
+            if (not json_error) and ('cause' in body):
+                message = self._get_error(body)
+            else:
+                message = body
+            raise ResourceConflictError(message, self.status)
+        else:
+            if (not json_error) and ('cause' in body):
```

Comment:
This logic seems to be repeated 4 times, perhaps it could be refactored inside a separate method (maybe even the existing ``_get_error`` method)?

Reply:
I originally grabbed most of this from the Google common lib.  Fixed this and the previous one - does it matter who clicks the 'resolve conversation' button (if it's used)?